sponge.go

```diff
@@ -45,57 +73,48 @@ func (h HackerNewsItem) getFormatted() Formatted {
 		Body: fmt.Sprintf("Title: %s\nUrl: %s", h.Title, h.Url)}
 }
 
-func getHackerNews() []Formatted {
+func getHackerNews() ([]Formatted, error) {
 	hackerNewsListUrl := "https://hacker-news.firebaseio.com/v0/topstories.json"
 	hackerNewsItemUrl := "https://hacker-news.firebaseio.com/v0/item/%d.json"
 
-	client := &http.Client{Timeout: 10 * time.Second}
-	resp, err := client.Get(hackerNewsListUrl)
-	if err != nil {
-		fmt.Println(err)
-		return make([]Formatted, 0)
-	}
-	defer resp.Body.Close()
-
-	hnl := make([]int, 0)
-	decoder := json.NewDecoder(resp.Body)
-	decoder.Decode(&hnl)
+	var hnl []int
+	getJsonResponse(hackerNewsListUrl, &hnl)
 
 	// take only top items (returns 500 initially)
-	hnTopItems := make([]int, itemsToFetch)
-	copy(hnTopItems, hnl[:itemsToFetch])
-
-	hnTopItemsDetails := make([]Formatted, 0)
+	var hnTopItems []int
+	hnTopItems = append(hnTopItems, hnl[:*itemsToFetch]...)
+	var hnTopItemsDetails []Formatted
+	collectorChan := make(chan Formatted)
 
 	var wg sync.WaitGroup
-	wg.Add(itemsToFetch)
+	wg.Add(len(hnTopItems))
```

Comment:
This is definitely the right way to do it here but sometimes you don't know in advance if you will want to run a goroutine for an item in advance.  It's also valid to `wg.Add(1)` and I do this a lot.  I do it by default, in fact, because I'm not always prepared to put the mental effort into whether a goroutine *will* run for each item and I'd rather not have a deadlock.

by wg.Add(N) is significantly faster for large lists and short lived goroutines.  The speed difference is quickly overshadowed though since it's basically just a lock, test, and add.